### ISSUE TITLE
Unify CoreLib/System.Runtime.Numerics formatting/parsing code

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#28657]
Once upon a time it seems as though System.Runtime.Numerics forked the integer formatting/parsing code that's in CoreLib.  That CoreLib code has since evolved, and is also now being mirrored to corefx, where System.Runtime.Numerics could include it rather than maintaining its own copy.

cc: @tannergooding

**Comment by ViktorHofer** [comment]
@stephentoub is that still the case with the consolidated repo?

**Comment by stephentoub** [comment]
Yup

**Comment by huoyaoyuan** [comment]
Looking at this. `Number` is using a lot of CoreLib internals. Though many of them isn't required by System.Runtime.Numerics, it isn't factored to be included only a part. Some others like FastAllocateString are critical to performance.
Folding it into CoreLib doesn't sound correct to me. @stephentoub any thoughts here?

**Comment by huoyaoyuan** [comment]
It does help a lot to fold into CoreLib, but it doesn't sound correct that *anything numeric related are required to be in CoreLib to get best performance*.

Currently blockers to compile `Number` out of CoreLib:
- FastAllocateString. How does `ValueStringBuilder` perform comparing to it?
- Decimal/Half manipulating. Can become public API, or use reinterpret cast.
- NumberFormatInfo internals. They are all available through public API, however those APIs has to defensive copy for array-returning members.
  - This also means that any third-party library can't utilize `NumberFormatInfo` with good performance

**Comment by stephentoub** [comment]
> anything numeric related are required to be in CoreLib to get best performance

It is likely always going to be the case that internal implementation details of corelib can be utilized to eek out the best perf.  For example, you mention FastAllocateString: that will never be exposed publicly, because it's an implementation detail of creating a mutable System.String that's written to and then published for immutable consumption.  The public version of that is `String.Create`, which enables the same functionality but in a safe, span-based manner, the primary overhead of which is an extra delegate invocation.

> it isn't factored to be included only a part

Work here could include refactoring to split the file into pieces compiled into both assemblies and pieces that are unique to one or the other.  It could also (if feasible) involve changing portions of the BigInteger code to sit on top of public API instead where that's possible functionally and without significantly impacting performance.

**Comment by huoyaoyuan** [comment]
Yep. After looking a bit more, the only required thing is `NumberFormatInfo`. It's API shape is unfriendly for performance, and the number formatting code access internal fields directly to avoid defensive copy. I think it's better to open a new issue for discussing it.
